qcodes/dataset/data_set.py

```diff
@@ -287,8 +288,10 @@ def __init__(self, path_to_db: Optional[str] = None,
             # of a new dataset. Note that a dataset is created (in the DB)
             # with no parameters; they are written to disk when the dataset
             # is marked as started
-            if exp_id is None:
-                exp_id = get_last_experiment(self.conn)
+            if exp_id is None:  # First, try to get the active exp_id
+                exp_id = get_active_experiment_id()
+                if exp_id is None:  # If no active exp_id, get the last exp_id
+                    exp_id = get_last_experiment(self.conn)
                 if exp_id is None:  # if it's still None, then...
                     raise ValueError("No experiments found."
```

Comment:
i think my only real comment is about this place in the code. what happens here is that DataSet's contructor becomes the one owning the logic that "if there's no active experiment, then use the max-exp-id one, and if that's not there, then raise and ask user to create an experiment" - my feeling is that it's not DataSet's responsibility to know that, but rather the responsibility of the "active experiment" feature. If "get_active_experiment" would implement these details then DataSet (and potentially any other objects) can rely on the active experiment as default, and what happens when active experiemtn is not defined is not the business of the object that uses it. so, a bit of responsibility decoupling.

what do you think @FarBo @jenshnielsen ? if the selected code was actually part of get_active_experiment_id?

Reply:
I don't think it should be part of that function necessarily but it would probably make sense to refactor it out next to that one as a higher level function.

Reply:
This is a fair point, and nice to decouple active_experiment from DataSet.
I will try to see what I can do :)